File: violet/lib/services/bookmark_sync.dart
import 'dart:convert';
import 'package:flutter/foundation.dart';
import 'package:http/http.dart' as http;
import 'package:shared_preferences/shared_preferences.dart';
import 'package:synchronized/synchronized.dart';
import 'package:uuid/uuid.dart';
import 'package:violet/database/user/user.dart';
import 'package:violet/log/log.dart';

class BookmarkSync {
  static final Lock _lock = Lock();
  static final ValueNotifier<int> changes = ValueNotifier(0);
  static final _numeric = RegExp(r'^\d{1,20}$');

  static Set<String> _ids(List<Map<String, Object?>> rows) => rows
      .map((row) => row['Article'].toString())
      .where((id) => _numeric.hasMatch(id))
      .toSet();

  static Future<void> automatic() async {
    try {
      await sync();
    } catch (_) {
      Logger.error('[BookmarkSync] Automatic sync failed; local data retained');
    }
  }

  static Future<int?> sync({bool force = false}) =>
      _lock.synchronized(() async {
        final prefs = await SharedPreferences.getInstance();
        final token = prefs.getString('bookmark_sync_token') ?? '';
        final baseUrl = prefs.getString('bookmark_sync_server') ?? '';
        if (token.isEmpty || baseUrl.isEmpty) return null;
        final endpoint =
            '${baseUrl.replaceFirst(RegExp(r"/+$"), "")}/api/bookmark-sync';
        final days = prefs.getInt('bookmark_sync_days') == 7 ? 7 : 1;
        final manager = await CommonUserDatabase.getInstance();
        await manager.checkOpen();
        final db = manager.db!;
        await db.execute(
          'CREATE TABLE IF NOT EXISTS BookmarkSyncMeta '
          '(Id INTEGER PRIMARY KEY, State TEXT NOT NULL)',
        );

        final state = await db.transaction<Map<String, dynamic>?>((txn) async {
          final rows = await txn.query(
            'BookmarkSyncMeta',
            where: 'Id=?',
            whereArgs: [1],
          );
          final saved = rows.isEmpty
              ? <String, dynamic>{
                  'revision': 0,
                  'shadow': <String>[],
                  'lastSuccess': 0,
                }
              : Map<String, dynamic>.from(
                  jsonDecode(rows.first['State'] as String) as Map,
                );
          final now = DateTime.now().millisecondsSinceEpoch;
          if (!force && now - (saved['lastSuccess'] as int) < days * 86400000)
            return null;
          if (saved['pending'] == null) {
            final captured = _ids(await txn.query('BookmarkArticle'));
            final shadow = Set<String>.from(saved['shadow'] as List);
            saved['pending'] = {
              'captured': captured.toList(),
              'payload': {
                'requestId': const Uuid().v4(),
                'baseRevision': saved['revision'],
                'add': captured.difference(shadow).toList(),
                'remove': shadow.difference(captured).toList(),
              },
            };
          }
          await txn.rawInsert(
            'INSERT OR REPLACE INTO BookmarkSyncMeta VALUES (?, ?)',
            [1, jsonEncode(saved)],
          );
          return saved;
        });
        if (state == null) return null;
        final pending = Map<String, dynamic>.from(state['pending'] as Map);
        final response = await http
            .post(
              Uri.parse(endpoint),
              headers: {
                'Content-Type': 'application/json',
                'X-Violet-Sync-Token': token,
              },
              body: jsonEncode(pending['payload']),
            )
            .timeout(const Duration(seconds: 60));
        if (response.statusCode != 200)
          throw Exception('Sync HTTP ${response.statusCode}');
        final data = jsonDecode(response.body) as Map<String, dynamic>;
        final revision = data['revision'];
        final articles = data['articles'];
        if (revision is! int ||
            revision < (state['revision'] as int) ||
            articles is! List ||
            articles.length > 100000 ||
            articles.any((id) => id is! String || !_numeric.hasMatch(id))) {
          throw Exception('Invalid sync response');
        }
        final remote = Set<String>.from(articles);
        await db.transaction((txn) async {
          final rows = await txn.query('BookmarkArticle');
          final current = _ids(rows);
          final captured = Set<String>.from(pending['captured'] as List);
          final wanted = {...remote, ...current.difference(captured)}
            ..removeAll(captured.difference(current));
          final groups = await txn.query('BookmarkGroup', orderBy: 'Id');
          if (groups.isEmpty) throw Exception('No bookmark folder exists');
          final group = groups.firstWhere(
            (g) => g['Name'] == 'violet_default',
            orElse: () => groups.first,
          );
          final groupId = group['Id'];
          final batch = txn.batch();
          for (final row in rows) {
            final id = row['Article'].toString();
            if (_numeric.hasMatch(id) && !wanted.contains(id)) {
              batch.delete(
                'BookmarkArticle',
                where: 'Id=?',
                whereArgs: [row['Id']],
              );
            }
          }
          for (final id in wanted.difference(current)) {
            batch.insert('BookmarkArticle', {
              'Article': id,
              'DateTime': DateTime.now().toString(),
              'GroupId': groupId,
            });
          }
          await batch.commit(noResult: true);
          await txn.rawInsert(
            'INSERT OR REPLACE INTO BookmarkSyncMeta VALUES (?, ?)',
            [
              1,
              jsonEncode({
                'revision': revision,
                'shadow': remote.toList(),
                'lastSuccess': DateTime.now().millisecondsSinceEpoch,
              }),
            ],
          );
        });
        changes.value++;
        return remote.length;
      });
}

File: violet-web/packages/frontend/src/services/bookmark-sync.ts
import { syncActivity } from './activity-sync';
import { getGroups, getBookmarkArticles } from '../api/bookmarks';
import { transactBookmarkSync } from './user-database';

const CONFIG_KEY = 'violet-bookmark-sync-config';
const ENDPOINT = `${window.location.protocol}//${window.location.hostname}:3002/api/bookmark-sync`;
type Config = { token: string; days: number };
type Payload = { requestId: string; baseRevision: number; add: string[]; remove: string[] };
type State = { Id: 'state'; revision: number; shadow: string[]; lastSuccess: number;
  leaseUntil: number; pending?: { payload: Payload; captured: string[] } };

export function getBookmarkSyncConfig(): Config {
  try {
    const value = JSON.parse(localStorage.getItem(CONFIG_KEY) ?? '{}');
    return { token: typeof value.token === 'string' ? value.token : '', days: value.days === 7 ? 7 : 1 };
  } catch { return { token: '', days: 1 }; }
}
export function saveBookmarkSyncConfig(config: Config) {
  localStorage.setItem(CONFIG_KEY, JSON.stringify(config));
}
function requestId() {
  const bytes = crypto.getRandomValues(new Uint8Array(16));
  bytes[6] = (bytes[6] & 15) | 64;
  bytes[8] = (bytes[8] & 63) | 128;
  const hex = Array.from(bytes, b => b.toString(16).padStart(2, '0')).join('');
  return `${hex.slice(0,8)}-${hex.slice(8,12)}-${hex.slice(12,16)}-${hex.slice(16,20)}-${hex.slice(20)}`;
}
function ids(articles: Array<{ Article: string }>) {
  return [...new Set(articles.map(a => String(a.Article)).filter(x => /^\d{1,20}$/.test(x)))];
}

export async function syncBookmarks(force = false): Promise<number | null> {
  const count = await syncArticleBookmarks(force);
  await syncActivity(force);
  return count;
}
async function syncArticleBookmarks(force: boolean): Promise<number | null> {
  const config = getBookmarkSyncConfig();
  if (!config.token) return null;
  await getGroups();
  await getBookmarkArticles(); // Finish legacy storage migration first.
  const state = await transactBookmarkSync<State | null>((articles, saved, _store, meta) => {
    const s: State = saved ?? { Id: 'state', revision: 0, shadow: [], lastSuccess: 0, leaseUntil: 0 };
    const now = Date.now();
    if (s.leaseUntil > now || (!force && now - s.lastSuccess < config.days * 86400000)) return null;
    if (!s.pending) {
      const captured = ids(articles);
      const current = new Set(captured), shadow = new Set(s.shadow);
      s.pending = { captured, payload: { requestId: requestId(), baseRevision: s.revision,
        add: captured.filter(id => !shadow.has(id)), remove: s.shadow.filter(id => !current.has(id)) } };
    }
    s.leaseUntil = now + 120000;
    meta.put(s);
    return s;
  });
  if (!state?.pending) return null;
  try {
    const response = await fetch(ENDPOINT, { method: 'POST',
      headers: { 'Content-Type': 'application/json', 'X-Violet-Sync-Token': config.token },
      body: JSON.stringify(state.pending.payload), signal: AbortSignal.timeout(60000) });
    if (!response.ok) throw new Error(`Sync HTTP ${response.status}`);
    const data = await response.json();
    if (!Number.isSafeInteger(data.revision) || data.revision < state.revision ||
        !Array.isArray(data.articles) || data.articles.length > 100000 ||
        data.articles.some((x: unknown) => typeof x !== 'string' || !/^\d{1,20}$/.test(x))) {
      throw new Error('Invalid sync response');
    }
    const group = (await getGroups()).find(g => g.Name === 'violet_default') ?? (await getGroups())[0];
    await transactBookmarkSync((articles, saved: State, store, meta) => {
      if (saved?.pending?.payload.requestId !== state.pending!.payload.requestId) throw new Error('Sync state changed');
      const captured = new Set(state.pending!.captured);
      const current = new Set(ids(articles));
      const wanted = new Set<string>(data.articles);
      // Keep local edits made while the request was in flight.
      for (const id of current) if (!captured.has(id)) wanted.add(id);
      for (const id of captured) if (!current.has(id)) wanted.delete(id);
      let nextId = articles.reduce((n, a) => Math.max(n, a.Id), 0) + 1;
      for (const article of articles) if (/^\d{1,20}$/.test(String(article.Article)) && !wanted.has(String(article.Article))) store.delete(article.Id);
      for (const id of wanted) if (!current.has(id)) store.put({ Id: nextId++, Article: id, GroupId: group.Id, DateTime: new Date().toISOString() });
      meta.put({ Id: 'state', revision: data.revision, shadow: data.articles, lastSuccess: Date.now(), leaseUntil: 0 });
    });
    window.dispatchEvent(new Event('violet-bookmarks-synced'));
    return data.articles.length;
  } catch (error) {
    await transactBookmarkSync((_articles, saved: State, _store, meta) => {
      if (saved?.pending?.payload.requestId === state.pending!.payload.requestId) meta.put({ ...saved, leaseUntil: 0 });
    });
    throw error;
  }
}
